robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+
+	// Register a `Resource` that generates weak dependencies. Specifically instance of
+	// this resource will depend on every `component` resource.
+	weakAPI := resource.NewAPI(uuid.NewString(), "component", "weak")
+	weakModel := resource.DefaultModelFamily.WithModel(utils.RandomAlphaString(5))
+	weak1Name := resource.NewName(weakAPI, "weak1")
+	resource.Register(
+		weakAPI,
+		weakModel,
+		resource.Registration[*someTypeWithWeakAndStrongDeps, *someTypeWithWeakAndStrongDepsConfig]{
+			Constructor: func(
+				ctx context.Context,
+				deps resource.Dependencies,
+				conf resource.Config,
+				logger logging.Logger,
+			) (*someTypeWithWeakAndStrongDeps, error) {
+				return &someTypeWithWeakAndStrongDeps{
+					Named:     conf.ResourceName().AsNamed(),
+					resources: deps,
+				}, nil
+			},
+			WeakDependencies: []resource.Matcher{resource.TypeMatcher{Type: resource.APITypeComponentName}},
+		})
+	defer resource.Deregister(weakAPI, weakModel)
+
+	// Start robot with two components and one resource with weak dependencies
+	// (that also has an explicit dependency on one of the components).
+	// The following scenario is expected:
+	// 1) The two bases will configure first in parallel (which will each bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
+	// 3) weak1 will configure.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Robot startup")
+	base1Name := base.Named("base1")
+	base2Name := base.Named("base2")
+	weakCfg1 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+			{
+				Name:  base2Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+		},
+	}
+	test.That(t, weakCfg1.Ensure(false, logger), test.ShouldBeNil)
+	robot := setupLocalRobot(t, context.Background(), &weakCfg1, logger)
+
+	res, err := robot.ResourceByName(weak1Name)
+	// The resource was found and all dependencies were properly resolved.
+	test.That(t, err, test.ShouldBeNil)
+	weak1, err := resource.AsType[*someTypeWithWeakAndStrongDeps](res)
+	test.That(t, err, test.ShouldBeNil)
+	// Assert that the weak dependency was tracked.
+	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
+	test.That(t, weak1.resources, test.ShouldContainKey, base1Name)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 1)
+
+	// Reconfigure base1
+	// The following scenario is expected:
+	// 1) The base1 will reconfigure (which will bump the logical clock).
```

Comment:
To 5.

Reply:
this is mostly correct

> // 2) Prior to calling `base1.Reconfigure` and subsequently `SwapResource` (which bumps the // logical clock), `completeConfig` is called. `completeConfig` checks to see if there is a // need to call `updateWeakDependents`. Either condition is sufficient for calling // `updateWeakDependents`:

Both conditions need to be true. Also, `completeConfig` is the function that calls `base1.Reconfigure` and subsequently `SwapResource`, so maybe just more accurate to say 

> Prior to calling `base1.Reconfigure` and subsequently `SwapResource` (which bumps the // logical clock), `completeConfig` checks...

3) and 4) are correct but it reads to me like they are being described as two separate steps but they happen at the same time. weak1 will not reconfigure because the resource did not have a diff in the new config and its explicit dependent's reconfiguration did not error or is newly built. We also do execute a "reconfiguration level" every time, even if nothing gets reconfigured on that level. We first separate resources into the different levels, then check each individual resource within whether it should reconfigure.

```
// 3) weak1 will be processed in a separate reconfiguration level but it will not reconfigure. For a resource to reconfigure, one of the following conditions should be met:
    - Resource config has a diff in the new robot config
    - An "explicit" dependent reconfigured and resulted in:
      -  an error return value
      - a new resource object being created ("newly built"). `base1` is reconfigured "in place".
```